modules/custom-status/custom-status.php

```diff
@@ -459,747 +466,806 @@ function post_admin_header() {
-	 * @since 0.7
-	 */
-	function handle_add_custom_status() {
+			// Validate and sanitize the form data
+			$status_name = isset( $_POST['status_name'] ) ? sanitize_text_field( trim( $_POST['status_name'] ) ) : '';
+			$status_slug = sanitize_title( $status_name );
+			$status_description = isset( $_POST['status_description'] ) ? stripslashes( wp_filter_nohtml_kses( trim( $_POST['status_description'] ) ) ) : '';
+
+			/**
+			 * Form validation
+			 * - Name is required and can't conflict with an existing name or slug
+			 * - Description is optional
+			 */
+			$_REQUEST['form-errors'] = array();
+			// Check if name field was filled in
+			if ( empty( $status_name ) ) {
+				$_REQUEST['form-errors']['name'] = __( 'Please enter a name for the status', 'edit-flow' );
+			}
+			// Check that the name isn't numeric
+			if ( 0 != (int) $status_name ) {
+				$_REQUEST['form-errors']['name'] = __( 'Please enter a valid, non-numeric name for the status.', 'edit-flow' );
+			}
+			// Check that the status name doesn't exceed 20 chars
+			if ( strlen( $status_name ) > 20 ) {
+				$_REQUEST['form-errors']['name'] = __( 'Status name cannot exceed 20 characters. Please try a shorter name.', 'edit-flow' );
+			}
+			// Check to make sure the status doesn't already exist as another term because otherwise we'd get a weird slug
+			if ( term_exists( $status_slug, self::taxonomy_key ) ) {
+				$_REQUEST['form-errors']['name'] = __( 'Status name conflicts with existing term. Please choose another.', 'edit-flow' );
+			}
+			// Check to make sure the name is not restricted
+			if ( $this->is_restricted_status( strtolower( $status_slug ) ) ) {
+				$_REQUEST['form-errors']['name'] = __( 'Status name is restricted. Please choose another name.', 'edit-flow' );
+			}
 
-		// Check that the current POST request is our POST request
-		if ( !isset( $_POST['submit'], $_GET['page'], $_POST['action'] )
-			|| $_GET['page'] != $this->module->settings_slug || $_POST['action'] != 'add-new' )
+			// If there were any form errors, kick out and return them
+			// phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotValidated
+			if ( count( $_REQUEST['form-errors'] ) ) {
+				$_REQUEST['error'] = 'form-error';
 				return;
+			}
 
-		if ( !wp_verify_nonce( $_POST['_wpnonce'], 'custom-status-add-nonce' ) )
-			wp_die( $this->module->messages['nonce-failed'] );
-
-		// Validate and sanitize the form data
-		$status_name = sanitize_text_field( trim( $_POST['status_name'] ) );
-		$status_slug = sanitize_title( $status_name );
-		$status_description = stripslashes( wp_filter_nohtml_kses( trim( $_POST['status_description'] ) ) );
+			// Try to add the status
+			$status_args = array(
+				'description' => $status_description,
+				'slug' => $status_slug,
+			);
+			$return = $this->add_custom_status( $status_name, $status_args );
+			if ( is_wp_error( $return ) ) {
+				wp_die( esc_textarea( __( 'Could not add status: ', 'edit-flow' ) . $return->get_error_message() ) );
```

Comment:
Couple suggestions:

1. We can use `esc_html()` here which should be more restrictive. `add_custom_status()` seems to primarily return potential errors from `wp_insert_term()`, which returns basic string error messages: https://github.com/WordPress/wordpress-develop/blob/6.5/src/wp-includes/taxonomy.php#L2399-L2766
2. Just for improvement's sake we could do a `sprintf()` instead of string concatenation to allow a translator to reorder where the error appears in this message.


```suggestion
			/* translators: %s: error message */
			wp_die( esc_html( sprintf( __( 'Could not add status: %s', 'edit-flow' ), $return->get_error_message() ) ) );
```

Reply:
Done